### DISCUSSION TITLE
scala.version property and scala version used on runtime

### DISCUSSION THREAD

**Discussion by azdrojowa123** [davidB/scala-maven-plugin#803] [Q&A]
Hello, 
I'm developing IntelliJ Scala Plugin and I want to improve the Maven support.
I have a question about the `scala.version` property. I noticed that if it the property is defined, and there is no scala library in the dependencies (even in transitive dependencies) then the scala code can be compiled, but cannot be executed. And it is compiled with with the exact version from the properties. So you somehow add the scala library to the classpath. 

Question: Why if you under the hood add the scala library for compilation it's not listed in the dependencies? 

The situation is changing when there is any scala library in the dependencies (even in transitive deps) no matter with which scope. So e.g. If I have transitive scala library dependency with the `test` scope, I can run the main class from the sources, and not just from tests. 
Such a thing could not be done when we only had the `scala.version` property, without any scala library in the dependencies. 

Question: If there is any scala library in the dependencies, no matter with which scope, then how do you choose which one should be used on runtime? (I particularly have a case in mind where there is no scala library with the compile scope)

Thank you in advance for your answer :)

**Comment by slandelle** [comment]
Disclaimer: I'm not the owner of this project and have only been a maintainer for a few years, so I lack all the history here.

First, thanks for your great work supporting this plugin in IntelliJ.

IMHO, lots of features in this plugin are either useless, nonsensical or overly complicated while having low value.
I really mean no offense to the original contributors here: the code base is 16 years old so technical debt is expected, Scala was way way way less stable back in the days and contributors were probably sometimes too enthusiastic about now features to add.

I think this `scala.version` property is one of them.

I'm not sure how feasible a version 2 of this plugin is, where we would drop lots of boilerplate.
I have limited bandwidth on this and I don't know who else could help with it.
If some volunteers chime in, it would be great.

In the meantime, I think you should restrain from supporting this option.

**Comment by azdrojowa123** [comment]
What do you mean by 

> restrain from supporting this option 

? 
As I wrote in my first message, if there is no `scala-library` in the dependencies, adding `scala.version` property allows to compile the scala code. If we don't take this into account in the Scala plugin, then the user experience when using the Scala plugin, and the native Maven, will be different. 

Also it is not entirely clear to me how it happens that when I add `scala-library` with the `test` scope I can run some code in production sources (as there is no other `scala-library` in the `compile` scope). Probably you must somehow reuse this library and change its scope (?) I'm just guessing. 
But anyway if I print the dependencies the `scala-library` is still displayed with the `test` scope.

**Comment by slandelle** [comment]
I'm saying that this `scala.version` option should be dropped and that no one should be using it.
The scala-maven-plugin should only rely on the version of the scala-library that's in the classpath.

**Comment by azdrojowa123** [comment]
Hmm okey, it would be a big simplification for us, but it's supported now and we cannot go beyond how the `scala-maven-plugin` plugin works.

**Comment by slandelle** [comment]
This option is purely optional. Personally I never use it. I would recommend not supporting it.

What would be nice would be that other community members would join in creating a list of features that must be kept and features that must be dropped so we can clean up and release a new major version.

**Comment by azdrojowa123** [comment]
It will be very helpful. Do you have a community discord, or all questions go directly to github? 

btw do you know maybe what is the behaviour when `scala.verison` is declared but there is some scala library in the dependencies with the non-compile scope (e.g. `test`)? Based on my observations, it allows to use it on runtime in production sources, so you have to somehow choose some scala library and adjust its scope.

**Comment by slandelle** [comment]
> Do you have a community discord, or all questions go directly to github?

No discord. And sadly very few active users.

> btw do you know maybe what is the behaviour when scala.version is declared but there is some scala library in the dependencies with the non-compile scope (e.g. test)? Based on my observations, it allows to use it on runtime in production sources, so you have to somehow choose some scala library and adjust its scope.

Honestly, no idea. This feature predates by far my joining here and I've never used it as it's really not the maven idiomatic way to magically tweak the classpath. And I suspect whatever the existing behavior is, it's broken.
